# Practical exercise - GO enrichment of expanded/contracted gene families

> This exercise is meant to be run locally on your laptop using RStudio.

Scenario:

You have sequenced, assembled and annotated your Mucoromycota genomes. Orthofinder was used to identify ortholog groups (gene families) and CAFE5 was used to identify a list expanding and contracting gene families. Your next task is to use GO enrichment to investigate what kind of genes are in these lists.

GO enrichment will be performed on the [g:Profiler website](https://biit.cs.ut.ee/gprofiler/gost). Since we are bringing our own genomes we will have to create a custom GMT.


Input:

* gff3 files for each genome containing GO annotations (must be downloaded)
* Hierarchical ortholog group table from orthofinder (in git repo)
* CAFE5 output (in git repo)


> To download the gff files from saga to your laptop:

```
mkdir downloads
rsync -va <USERNAME>@saga.sigma2.no:/cluster/projects/nn9984k/data/mucoromycota_gff3_files downloads
```

## Step 1: Extract GO terms from GFF files

The GO annotations we are going to use were generated with funannotate and are contained in gff files.

Example exerpt of gff file (generated by funannotate):
```
contig_1	funannotate	gene	514378	515130	.	-	.	ID=GZUMRA1EN_000194;
contig_1	funannotate	mRNA	514378	515130	.	-	.	ID=GZUMRA1EN_000194-T1;Parent=GZUMRA1EN_000194;product=hypothetical protein;Ontology_term=GO:0003677,GO:0110165,GO:0043227,GO:0005829,GO:0043229,GO:0005737,GO:0016020,GO:0043226,GO:0005634,GO:0043231,GO:0005622;Dbxref=InterPro:IPR002836,InterPro:IPR036883,PFAM:PF01984;note=EggNog:ENOG503P69I,COG:D,BUSCO:28106at1913637;
contig_1	funannotate	exon	515068	515130	.	-	.	ID=GZUMRA1EN_000194-T1.exon1;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514916	514998	.	-	.	ID=GZUMRA1EN_000194-T1.exon2;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514739	514803	.	-	.	ID=GZUMRA1EN_000194-T1.exon3;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514502	514671	.	-	.	ID=GZUMRA1EN_000194-T1.exon4;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514378	514428	.	-	.	ID=GZUMRA1EN_000194-T1.exon5;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	515068	515130	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514916	514998	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514739	514803	.	-	1	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514502	514671	.	-	2	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514378	514428	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
```

We only need the GO terms and the gene IDs they are assigned to.

The GO annotations are in the mRNA lines somewhere in the last column (you will need scroll to right in the above textbox to see them there). The format is like this: `Ontology_term=GO:0003677,GO:0110165,...,GO:0005622;`. I.e. it always follows "Ontology_term=" and terms are separated by "," and ends with ";". The gene ID ca be found in the "Parent" of the mRNA.

The following R script uses regular expressions to extract the gene IDs and GO terms:

```
library(tidyverse)

# extract GO terms from GFF file
GOfromGFF <- function(gff_file){
  # Read GFF3 as tsv file. Only read the 9th column that contains "attributes"
  read_tsv(gff_file, col_types = "--------c",col_names = "attributes",comment = "#") %>% 
    filter(grepl("Ontology_term",attributes)) %>% # only keep entries with GO term
    mutate( gene_id = str_extract(attributes, "(?<=Parent=)[^;]+")) %>% 
    mutate( GO_id = str_extract(attributes, "(?<=Ontology_term=)[^;]+")) %>% 
    select( -attributes ) %>% 
    mutate( GO_id = strsplit(split = ",",GO_id)) %>%
    unnest( GO_id ) %>% 
    distinct()
}
```


> Optional: If you want, here is an example how to get GO annotations for a given gff and save as tsv file:

```
GOtbl <- GOfromGFF(gff_file = "downloads/mucoromycota_gff3_files/ggMorZona1.gff3")
write_tsv(GOtbl, file = "ggMorZona1_GO_annotations.tsv")
```


We will however not store each GO terms for each individual genome but rather combine them by gene family. But we still need to load the GO terms from all the gff files:

```
GOallGenomes <-
  dir("downloads/mucoromycota_gff3_files",pattern=".gff3$",full.names = T) %>% 
  setNames(.,sub(".gff3$","",basename(.))) %>% 
  lapply(GOfromGFF)
```


## Step 2: Combine GO terms per gene family

We used the N0 (root node) hierarchical orthogroups (HOGs) for running CAFE5 and want to assign the GO terms to each gene family.

If you have cloned this git repo you will find the file under the `data` folder and can load it in R like this:

```
N0_HOGs <- read_tsv("data/orthofinder/Phylogenetic_Hierarchical_Orthogroups/N0.tsv")
```

Following is the script to

```
HOG_GOtbl <- 
  N0_HOGs %>% 
  select(-OG,-`Gene Tree Parent Clade`) %>% 
  pivot_longer(cols = -HOG, names_to = "species", values_to = "gene_id") %>% 
  na.omit() %>% 
  mutate( gene_id = strsplit(split = ", ",gene_id)) %>%
  unnest( gene_id ) %>% 
  mutate( species = sub("\\..*$","",species)) %>% # remove suffix after .
  inner_join( bind_rows(GOallGenomes,.id="species"), by=c("species","gene_id")) %>% 
  distinct(HOG,GO_id)
```

Save this to a file:

```
write_tsv(HOG_GOtbl, "HOG_GOtbl.tsv")
```

## Step 3: Using GMT helper to convert to valid GMT file


There are many available tools for performing GO enrichment analysis. In this tutorial we will use [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost) which is easy to use and allows for uploading custom annotation with GMT files.

We currently have our custom GO annotations in tabular format and need to convert these to GMT format to use them in g:Profiler. 

> Note: The gmt format is a tab delimited format for describing gene sets. Each row is one gene set with first column being the gene set identifier (in our case GO term ids), second column containing a description (GO term description) and the following columns are the gene identifiers (ther can be different number of columns in each row). A description of the gmt format can be found [here](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29)

### Step 3.1: Convert tabular file to GMT

Go to [gmt-helper](https://biit.cs.ut.ee/gmt-helper/) and click the "Convert tabular file to GMT" tab. Click the button to upload your TSV file (You should now see the first couple of rows in the file). Select the column with the GO IDs as the "annotation column" and the gene IDs as the "ID column". Leave the "description column" empty.

When you click "Convert" you should get a GMT file names "result_*.gmt". It is a good idea to rename this to something more meaningful.

### Step 3.2: Reannotate GMT using OBO

It is common to only annotate genes with most specific term in the GO hierarchy while the the more generic terms left out as they can be inferred by propagating the annotations up the GO hierarchy. When using a GMT file in g:Profiler it does not automatically propagate the annotations so we need to use the GMT helper again for this.

* To do this we first need to download the `go-basic.obo` file from (https://geneontology.org/docs/download-ontology/) which contains the full GO hierarchy.
* Then go back to the GMT helper and open the "Reannotate GMT using OBO" tab
* Upload your GMT file and the OBO file
* Leave the selected relations at "is_a" and "part_of" and click "Reannotate"

After some seconds you should get a new GMT file with propagated GO terms. The GO terms will now also have descriptions.

## Step 4: Get list

```
significant_family <-read_tsv("data/cafe/cafe_results/Base_family_results.txt") %>%
  filter(`Significant at 0.05`=='y')

write_lines(changes_significant$FamilyID,file = "CAFE5_significant_families.txt")
```

## Step 5: Run enrichment...

## Step 6: Analyse results

> You can copy the significant terms and view them in [QuickGO](https://www.ebi.ac.uk/QuickGO/) by adding them to basket...

**To be continued...**
 
