# Practical exercise - GO enrichment of expanded/contracted gene families

Scenario:

You have sequenced, assembled and annotated your Mucoromycota genomes. Orthofinder was used to identify ortholog groups (gene families) and CAFE5 was used to identify a list expanding and contracting gene families. Your next task is to investigate what kind of genes are in these lists.

Input:

* gff3 files for each genome containing GO annotations (must be downloaded)
* Hierarchical ortholog group table from orthofinder (in git repo)
* CAFE5 output (in git repo)


> To download the gff files:

```
mkdir downloads
rsync -va <USERNAME>@saga.sigma2.no:/cluster/projects/nn9984k/data/mucoromycota_gff3_files downloads
```

## Step 1: Extract GO terms from GFF files

The GO annotations we are going to use were generated with funannotate and are contained in gff files.

Example gff file (generated by funannotate):
```
contig_1	funannotate	gene	514378	515130	.	-	.	ID=GZUMRA1EN_000194;
contig_1	funannotate	mRNA	514378	515130	.	-	.	ID=GZUMRA1EN_000194-T1;Parent=GZUMRA1EN_000194;product=hypothetical protein;Ontology_term=GO:0003677,GO:0110165,GO:0043227,GO:0005829,GO:0043229,GO:0005737,GO:0016020,GO:0043226,GO:0005634,GO:0043231,GO:0005622;Dbxref=InterPro:IPR002836,InterPro:IPR036883,PFAM:PF01984;note=EggNog:ENOG503P69I,COG:D,BUSCO:28106at1913637;
contig_1	funannotate	exon	515068	515130	.	-	.	ID=GZUMRA1EN_000194-T1.exon1;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514916	514998	.	-	.	ID=GZUMRA1EN_000194-T1.exon2;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514739	514803	.	-	.	ID=GZUMRA1EN_000194-T1.exon3;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514502	514671	.	-	.	ID=GZUMRA1EN_000194-T1.exon4;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	exon	514378	514428	.	-	.	ID=GZUMRA1EN_000194-T1.exon5;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	515068	515130	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514916	514998	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514739	514803	.	-	1	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514502	514671	.	-	2	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
contig_1	funannotate	CDS	514378	514428	.	-	0	ID=GZUMRA1EN_000194-T1.cds;Parent=GZUMRA1EN_000194-T1;
```

We only need the GO terms and the gene IDs they are assigned to.

The GO annotations are in the mRNA lines somewhere in the last column (you will need scroll to right in the above textbox to see them there). The format is like this: `Ontology_term=GO:0003677,GO:0110165,...,GO:0005622;`. I.e. it always follows "Ontology_term=" and terms are separated by "," and ends with ";". The gene ID ca be found in the "Parent" of the mRNA.

The following R script uses regular expressions to extract the gene IDs and GO terms:

```
library(tidyverse)

# extract GO terms from GFF file
GOfromGFF <- function(gff_file){
  # Read GFF3 as tsv file. Only read the 9th column that contains "attributes"
  read_tsv(gff_file, col_types = "--------c",col_names = "attributes",comment = "#") %>% 
    filter(grepl("Ontology_term",attributes)) %>% # only keep entries with GO term
    mutate( gene_id = str_extract(attributes, "(?<=Parent=)[^;]+")) %>% 
    mutate( GO_id = str_extract(attributes, "(?<=Ontology_term=)[^;]+")) %>% 
    select( -attributes ) %>% 
    mutate( GO_id = strsplit(split = ",",GO_id)) %>%
    unnest( GO_id ) %>% 
    distinct()
}
```


 `data/orthofinder/Phylogenetic_Hierarchical_Orthogroups/N0.tsv`

 
