---
title: "Comparative Genomics: Plotting Orthofinder Results"
author: "Helle Tessand Baalsrud"
date: "`r Sys.Date()`"
params:
  load.results: false
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/hellebaalsrud/Documents/Projects/Mucormycota/Orthofinder/Results_Mar18/")
knitr::opts_chunk$set(echo = TRUE)
load.results <- params$load.results
```

## Overview

The goal of this exercise is to plot some of the statistics given by orthofinder on the phylogenetic species tree, also generated by orthofinder. An example of this can be found in the [orthofinder publication](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1832-y), see figure below. 

![](/Users/hellebaalsrud/Desktop/orthofinder.png)    
# Installing R packages

First you should make sure you have all the libraries you will need for this exercise. You may have to install some of the packages. 

```{r, message=FALSE, warning=FALSE}
# to install packages, remove the hashtags:
  #install.packages("ggplot2")
  #install.packages("tibble")
  #install.packages("tidyr")
  #install.packages("ggstance")
  #install.packages("patchwork")
  #install.packages("aplot")
  #install.packages("ape")
  #install.packages("tidyverse")

  #if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
  #BiocManager::install("ggtree")

#libraries
library(tibble)
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggstance)
library(ggtree)
library(patchwork)
library(aplot)
library(ape)
library(tidyverse)

```

# Parsing output data using Bash

Orthofinder outputs a lot of tsv files. We are interested in plotting some of the data from the file "Comparative_Genomics_Statistics/Statistics_Per_Species". If you open up this file using Excel or Numbers it should look like this: 

![](/Users/hellebaalsrud/Desktop/stat_per_species.png)    


There are different ways of parsing this data so that it is more practical to analyse it in R. This is only one way of going it, you can also do this using python etc. 

Here we will use bash to solve the problem. You can either open up a terminal window and run this code there. Or you can add a code chunk in RStudio, and select bash instead of R:

![](/Users/hellebaalsrud/Desktop/bash.png)    

```{bash}
# if you need to downliad the data to your local computer first, you can run rsync.
  # rsync -avz username@fox... /Path_to_/Orthofinder/
  
  # parse excel tsv file from orthofinder:
  # we only want the first 11 lines in the file
sed -n '1,11 p' Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv >         Comparative_Genomics_Statistics/Statistics_PerSpecies_copy.tsv
  #then we will delete the first column
awk 'BEGIN { FS="\t";OFS="\t" } { $1=""; print$0  }' Comparative_Genomics_Statistics/Statistics_PerSpecies_copy.tsv > Comparative_Genomics_Statistics/Statistics_PerSpecies_copy2.tsv

  # NB! Here I have made copies of the files, to not overwrite the original file. You can of course overwrite them if you do not need the manymore. 
```

# Plotting your data in R

You should set your working directory to the folder where your results are. Change the code below and enter:

```{r}
#setwd("/Your_path_/Orthofinder/Results_Dec01_1/")
```

NB! If you are using R markdown (.Rmd) you need to change the working directory for the entire documnet, not just each code chunk. Then you need to add a code chung in the beginning that looks like this (change to your local path): 

![](/Users/hellebaalsrud/Desktop/setwd.png) 


Now we are ready to import our data into R: 

```{r, message=FALSE, warning=FALSE}
# import tsv file to R

dff<-read.table("Comparative_Genomics_Statistics/Statistics_PerSpecies_copy2.tsv",sep="\t")

# need to clean up a but and remove the first column:
dff[1] <- NULL 


#Then we want to flip the data around using transpose: 
df <- t(dff)
str(df)
#df <- data.frame(Species = row.names(df), df)
df <- data.frame(df)
# Here we will add column names that are suitable for R (No spaces): 

colnames(df) = c("Species", "Number.genes", "Number.genes.in.orthogroups", "Number.unassigned.genes", "Perc.genes.orthogroups", "Perc.uassigned.genes", "Number.orthogroups.containing.species", "Perc.orthogroups.containing.species", "Number.species.specific.orthogroups", "Number.genes.species.specific.orthogroups", "Perc.genes.species.specific.orthogroups" )

# we can also extract the name of the species and type of data from the name
df <-  df %>% separate(Species, c('Species2', 'Species_data'),remove = FALSE)

df <- df %>% 
    mutate_at(vars(Number.genes, Number.genes.in.orthogroups, Number.unassigned.genes,Perc.genes.orthogroups, Perc.uassigned.genes, Perc.orthogroups.containing.species, Perc.genes.species.specific.orthogroups,  Number.species.specific.orthogroups , Number.genes.species.specific.orthogroups,Perc.genes.species.specific.orthogroups, Number.orthogroups.containing.species), as.numeric)

```

We also need a phylogenetic tree! Our tree can be found in the folder "Species_Tree". 


```{r, message=FALSE, warning=FALSE}
#load tree
getwd()
tree <- read.tree("Species_Tree/SpeciesTree_rooted.txt")

tree_plot <- ggtree(tree) 
tree_plot + geom_tiplab() + ggplot2::xlim(0, 2)

```

Maybe we want to make the tip labels a bit nicer? We can use the package "ape" to do this. However, in the final plot we will use the labels from the dataframe, so for now it is OK that the tree has ugly names. 

```{r, message=FALSE, warning=FALSE}
# Our tree is a list
str(tree)

# we can examine the tip labels like this
tree[["tip.label"]]

```


Now we want to add some traits to the tree. We could start by plotting the number of genes: 

```{r, message=FALSE, warning=FALSE}
trait <- ggplot(df, aes(Number.genes, Species, label= Species)) + geom_colh(colour = "lemonchiffon3", fill="lemonchiffon3") + 
  scale_y_discrete(position = "right") +
  xlab("Number of genes") +
   ylab("Species") +
  theme_minimal()
trait
```

To combine the column plot we have just generated with our tree we will use the package "aplot". This online [book](https://yulab-smu.top/treedata-book/chapter7.html) has a lot of great examples of plotting trees. 

```{r, message=FALSE, warning=FALSE}
trait %>% insert_left(tree_plot, width=0.3)
```

Now this is starting to look like something! Lets add more traits, more plots!

Notice that the order of the species is different between the plots? This is because aplot aligns plots based on the order if species in the phylogenetic tree, instead of simply pasting them next to each other. Therefore it is essential that the tiplabels matches the names in the plot you are trying to align to the tree. 


```{r, message=FALSE, warning=FALSE}
trait <- ggplot(df, aes(Number.genes, Species, label= Species)) + geom_colh(colour = "lemonchiffon3", fill="lemonchiffon3") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("Number genes") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1), 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 


trait2 <- ggplot(df, aes(Perc.genes.orthogroups, Species, label= Species)) + geom_colh(colour = "aquamarine3", fill="aquamarine3") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("% genes OGs") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1),
        axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 
  

trait3 <- ggplot(df, aes(Number.species.specific.orthogroups, Species, label= Species2)) + geom_colh(colour = "coral", fill="coral") + 
  #scale_x_continuous(expand=c(0,0)) +
  scale_y_discrete(position = "right") +
  # scale_y_discrete(position = "right", labels=df$Species) +
  xlab("Species-specific OGs") +
   ylab("Species") +
   theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1)) 
 
trait4 <- ggplot(df, aes(Perc.orthogroups.containing.species, Species, label= NULL)) + geom_colh(colour = "cornflowerblue", fill="cornflowerblue") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("% OGs with species") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1), 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 


trait %>% insert_left(tree_plot, width=1.5)%>% insert_right(trait2) %>% insert_right(trait4) %>% insert_right(trait3) 

```

### Adding gene duplication events

In another results folder, there is another species tree file with inferred gene duplication events.

```{r, message=FALSE, warning=FALSE}
tree2 <- read.tree("Gene_Duplication_Events/SpeciesTree_Gene_Duplications_0.5_Support.txt")

tree_plot2 <- ggtree(tree2) 
tree_plot2 + geom_tiplab() + ggplot2::xlim(0, 2) + geom_nodelab(hjust=.5)
  
```
The number of duplications are embedded in the nodel labels and tiplabels. We can inspect them like so: 

```{r}
tree2[["tip.label"]]
tree2[["node.label"]]
```
```{r, message=FALSE, warning=FALSE}
#make dataframes for tiplabels and nodelabels. 
  #Using this you could basically add any annotation you would like to your tree plot. 
tips<-  data.frame(tree2[["tip.label"]]) %>% separate(tree2...tip.label..., c('TP', 'Tip_dup'),sep="_",remove = FALSE) %>% separate(TP, c('Tip','type'),remove =FALSE)
colnames(tips)[1] <- "Species"

nodes<-  data.frame(tree2[["node.label"]]) %>% separate(tree2...node.label..., c('VP', 'Node_dup'),sep="_",remove = FALSE) %>% separate(VP, c('Node','type'),remove =FALSE)

#combines the data with the tree plot object.  
d <- tree_plot2 %<+% tips %<+% nodes

tree_plot_dup<- d +  
  geom_text(aes(label=Node_dup), hjust=-.3, vjust=1.1)+
  geom_tiplab(aes(label=Tip_dup), 
                    align=T,
                    linetype = "dotted", linesize = .7) +
  ggplot2::xlim(0, 1.1)

tree_plot_dup


```

We now have to add the new tiplabels to our dataframe to be able to align the new tree with our orthofinder output data. 

```{r, message=FALSE, warning=FALSE}
df <-  merge(df, tips, by.x = "Species2", by.y = "Tip", all.x=T)
#lets change the name of this column so that it matches the name of the column from the plots for each trait:
colnames(df)[14] <- "Species"
```



```{r, message=FALSE, warning=FALSE}
trait <- ggplot(df, aes(Number.genes, Species, label= Species)) + geom_colh(colour = "lemonchiffon3", fill="lemonchiffon3") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("Number genes") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1), 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 


trait2 <- ggplot(df, aes(Perc.genes.orthogroups, Species, label= NULL)) + geom_colh(colour = "aquamarine3", fill="aquamarine3") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("% genes OGs") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1),
        axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 
  

trait3 <- ggplot(df, aes(Number.species.specific.orthogroups, Species, label= Species2)) + geom_colh(colour = "coral", fill="coral") + 
  #scale_x_continuous(expand=c(0,0)) +
  scale_y_discrete(position = "right") +
  #scale_y_discrete(position = "right", labels=df$Species2) +
  xlab("Species-specific OGs") +
   ylab("Species") +
   theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1)) 
 
trait4 <- ggplot(df, aes(Perc.orthogroups.containing.species, Species, label= Species2)) + geom_colh(colour = "cornflowerblue", fill="cornflowerblue") + 
  #scale_x_continuous(expand=c(0,0)) +
   scale_y_discrete(position = "right") +
  ylab(NULL) +
  xlab("% OGs with species") +
  theme_minimal() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1), 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()) 


trait %>% insert_left(tree_plot_dup, width=1.8)%>% insert_right(trait2) %>% insert_right(trait4) %>% insert_right(trait3) 


```

If you want to save your plot you can use ggsave:

```{r}
#ggsave("/name_of_file.pdf", width = 18, height = 18, units = "cm")

```


